src/data/nat/factorization.lean

```diff
@@ -119,6 +120,32 @@ begin
     simp [prod_insert hxT, sum_insert hxT, ←IH, factorization_mul (hS x hxS) hT] }
 end
 
+/-! ### Bijection between pnats and finsupps `ℕ →₀ ℕ` with support on the primes -/
+
+/-- Any finsupp `f : ℕ →₀ ℕ` whose support is in the primes is equal to the factorization of
+the product `∏ (a : ℕ) in f.support, a ^ f a`. -/
+lemma factorization_prod_pow_inv {f : ℕ →₀ ℕ} (hf : ∀ (p : ℕ), p ∈ f.support → prime p) :
+  (f.prod pow).factorization = f :=
+begin
+  have h : ∀ x : ℕ, x ∈ f.support → x ^ f x ≠ 0 := λ p hp, pow_ne_zero _ (prime.ne_zero (hf p hp)),
+  simp only [finsupp.prod, factorization_prod h],
+  nth_rewrite_rhs 0 (sum_single f).symm,
+  exact sum_congr rfl (λ p hp, prime.factorization_pow (hf p hp)),
+end
+
+lemma eq_factorization_iff {n : ℕ} {f : ℕ →₀ ℕ} (hn : n ≠ 0) (hf : ∀ p ∈ f.support, prime p) :
+  f.prod pow = n ↔ f = n.factorization :=
+⟨λ h, by rw [←h, factorization_prod_pow_inv hf],
+ λ h, by rw [h, factorization_prod_pow_eq_self hn]⟩
+
+/-- The positive natural numbers are bijective with finsupps `ℕ →₀ ℕ` with support in the primes -/
+noncomputable
+def factorization_equiv : ℕ+ ≃ {f : ℕ →₀ ℕ | ∀ p ∈ f.support, prime p} :=
```

Comment:
Can you add the relevant `apply` and `inv_apply` lemmas? I think `@[simps]` can produce them automatically.

Reply:
I've added `@[simps]`, which generates
```
factorization_equiv_apply : ∀ (_x : ℕ+), factorization_equiv _x = factorization_equiv._match_1 _x
```
```
factorization_equiv_symm_apply : ∀ (_x : ↥{f : ℕ →₀ ℕ | ∀ (p : ℕ), p ∈ f.support → prime p}),
    (factorization_equiv.symm) _x = factorization_equiv._match_2 _x
```
but these don't seem very useful.  Is there a nicer form of these that you were thinking of that I might prove directly?

Reply:
I agree it's not very nice. You can prove something like `(factorization_equiv n).1 = n.1.factorization`, the proof is `rfl` I suppose. Similarly for `inv_apply`.